src/v/model/record.h

```diff
@@ -350,6 +350,17 @@ class record_batch_attributes final {
     std::bitset<16> _attributes;
 };
 
+struct key_offset {
+    key_offset(int32_t od, int32_t ks, bytes&& k)
+      : offset_delta(od)
+      , key_size(ks)
+      , key(std::move(k)) {}
+
+    int32_t offset_delta;
+    int32_t key_size;
+    bytes key;
```

Comment:
this seems very expensive. This struct is 48 bytes, per key. so a batch of 1K records is now over a full page. 

I think you can store this on a simpler struct. 

```
struct record_batch_key_offset {

    int32_t offset_delta; 
    int32_t key_size;  ( alex is not sure why we need the key_size)
    int64_t key_hash;
};

```

Is there a world where we use the key_hash - say xxhash64 or some other good non crypto hash that degrades into a *secondary batch decompression* if and only if there is a collision?

Reply:
We copy a batch [here](https://github.com/vectorizedio/redpanda/pull/2435/commits/517a4ad490870491981e11728059c43260d1195d#diff-057947c463347b74bf5eef62d2b15f6efcc2ff244c53ac3f93892a4a9ff0546fR190) but avoid copying it [there](https://github.com/vectorizedio/redpanda/blob/076105985c9ddf2d7056350907da5a30cfaa34c9/src/v/model/record_batch_reader.h#L115) by switching from make_foreign_memory_record_batch_reader to make_memory_record_batch_reader

We used to have a compressed/decompressed batch at the same time [there](https://github.com/vectorizedio/redpanda/pull/2435/commits/99eb85f3f211d6651d18beee42718757f8862fcc#diff-7a75bec80b6c419b49048c0bcc83a6d08bdf9f02288789393a4c2794dfb3ed15R366) and now we decompress and cache it [here](https://github.com/vectorizedio/redpanda/pull/2435/commits/517a4ad490870491981e11728059c43260d1195d#diff-057947c463347b74bf5eef62d2b15f6efcc2ff244c53ac3f93892a4a9ff0546fR187) so memory pressure looks the same.